[user triton] triton_kernel_wrap support for new host-side TMA API

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155896
* #155879
* #155827
* #155814
* __->__ #155777

This adds support for user-defined triton kernels using TensorDescriptor.from_tensor into triton_kernel_wrap: i.e. storing metadata about the TMA descriptors and doing mutation analysis.

Major changes:
* TMADescriptorMetadata has changed: previously it was a dict[str, tuple[list[int], list[int], int]]. But now there are two metadata formats: one for experimental API and one for stable API. Now the metadata format is dict[str, tuple[str, tuple[...]]], where tuple[...] is tuple[list[int], list[int], int] for experimental and tuple[list[int],] for stable API. And then most handling of the metadata has to be branched based on whether the metadata represents a stable or experimental TMA descriptor
* mutation analysis: unlike experimental TMA (where the mutation analysis / ttir analysis pretends that the TMA descriptor is actually just a tensor), we need to construct an actual TMA descriptor before getting the Triton frontend to create the TTIR (otherwise assertions fail). A TensorDescriptor (i.e. stable TMA API descriptor) passed into a python triton kernel actually turns into 1 + 2*N parameters in the TTIR (for a rank-N tensor), so the arg list also needs to be patched for this reason (in generate_ttir)
* mutation analysis: now we also need to pass tma_descriptor_metadata into the mutation analysis, in order to create the TMA descriptors that are passed into the frontend code (ie. the previous point). This is why all the mutation tests are modified with an extra return value (the tma_descriptor_metadata)

Inductor is not modified (Inductor just errors out if you use a stable API tma descriptor). This will be the next PR.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

davidberard98 committed:
[user triton] triton_kernel_wrap support for new host-side TMA API

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] triton_kernel_wrap support for new host-side TMA API"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update on "[user triton] triton_kernel_wrap support for new host-side TMA API"


This adds support for user-defined triton kernels using TensorDescriptor.from_tensor into triton_kernel_wrap: i.e. storing metadata about the TMA descriptors and doing mutation analysis.

Major changes:
* TMADescriptorMetadata has changed: previously it was a dict[str, tuple[list[int], list[int], int]]. But now there are two metadata formats: one for experimental API and one for stable API. Now the metadata format is dict[str, tuple[str, tuple[...]]], where tuple[...] is tuple[list[int], list[int], int] for experimental and tuple[list[int],] for stable API. And then most handling of the metadata has to be branched based on whether the metadata represents a stable or experimental TMA descriptor
* mutation analysis: unlike experimental TMA (where the mutation analysis / ttir analysis pretends that the TMA descriptor is actually just a tensor), we need to construct an actual TMA descriptor before getting the Triton frontend to create the TTIR (otherwise assertions fail). A TensorDescriptor (i.e. stable TMA API descriptor) passed into a python triton kernel actually turns into 1 + 2*N parameters in the TTIR (for a rank-N tensor), so the arg list also needs to be patched for this reason (in generate_ttir)
* mutation analysis: now we also need to pass tma_descriptor_metadata into the mutation analysis, in order to create the TMA descriptors that are passed into the frontend code (ie. the previous point). This is why all the mutation tests are modified with an extra return value (the tma_descriptor_metadata)

Inductor is not modified (Inductor just errors out if you use a stable API tma descriptor). This will be the next PR.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

davidberard98 committed:
Update

[ghstack-poisoned]

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -231,15 +288,69 @@ def generate_ttir(
-    ordered_tensor_names = [
-        name for name, arg in ordered_args.items() if isinstance(arg, Tensor)
-    ]
+    def is_stable_tensor_descriptor_arg(arg: Any) -> bool:
```
Should we cache on this? Can this happen repetitively on a single arg during the analysis?

davidberard98 replied:
@aakhundov I think this check should happen infrequently: it'll happen a constant number of times for every arg passed into a triton kernel: so this should be relatively cheap compared to the other analysis (which happens for every op in the graph)

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -720,6 +831,7 @@ def analyze_kernel_mutations(
         "tt.atomic_rmw": [0],
         "tt.experimental_descriptor_store": [0],
         "tt.experimental_tensormap_create": [0],
+        "tt.descriptor_store": [0],
```
How are the mutated TMA descriptor args associated with the mutation of the underlying tensor in the fx graph?

davidberard98 replied:
It works the same way as the old TMA API (I had been worried about this part, but it turns out it all works out the same as before)

1. the fx graph for the triton kernel takes in the underlying tensor as an argument (and also stores some metadata about the tma descriptor):

```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 3, grid = [(2, 1, 1)], tma_descriptor_metadata = {'in_desc_ptr0': ('stable', ([256],)), 'in_desc_ptr1': ('stable', ([256],)), 'out_desc_ptr': ('stable', ([256],))}, kwargs = {'in_desc_ptr0': arg0_1, 'in_desc_ptr1': arg1_1, 'out_desc_ptr': full_default}, tensors_to_clone = ['out_desc_ptr']);  arg0_1 = arg1_1 = full_default = None
```

2. the generate_ttir step realizes this is actually a TMA descriptor, so it constructs a fake TMA descriptor for the TTIR creation and adds the fake inputs as parameters. Specifically, the **tensor input name is mapped to the TMA descriptor in the TTIR** so mutations that occur to the TMA descriptor in the TTIR will actually register as mutations to the underlying tensor. TL;DR: it's all the same as the old TMA API, except we have to do a bit of extra work in the TTIR processing to account for the additional stride/shape parameters.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -231,15 +288,69 @@ def generate_ttir(
+            block_shape = stable_meta[0]
+            tensor_rank = len(block_shape)
+            names = [name]
+            names.extend(name + f" STRIDE PLACEHOLDER {i}" for i in range(tensor_rank))
+            names.extend(name + f" SIZE PLACEHOLDER {i}" for i in range(tensor_rank))
```
This is so bad. Why is this done for the host-side TMA? Is the information not already in the TMA descriptor prefilled before launching the kernel?

davidberard98 replied:
IIUC, it's because the Triton kernel may need to use this information (e.g. outside of the direct TMA loads/stores, you may want to know the stride/block sizes), but they don't want to reverse-engineer the TMA descriptor format to be able to read this data - so instead they manually pass in the values individually

see the comment here: https://github.com/triton-lang/triton/blob/main/third_party/nvidia/backend/driver.py#L129-L133

aakhundov commented on torch/_inductor/ir.py:
```diff
@@ -6077,13 +6077,23 @@ class TMADescriptor(ExternKernel):
+        exp_meta = maybe_unpack_tma_experimental_metadata(tma_meta)
+        if exp_meta is None:
+            raise NotImplementedError(
+                f"tma_meta {tma_meta} not recognized (if this kernel uses TensorDescriptor.from_tensor, "
```
Is `TensorDescriptor.from_tensor` actually used in the kernel?

davidberard98 replied:
good point, I'll clarify this as "if this kernel uses a TMA descriptor created with TensorDescriptor.from_tensor..."

aakhundov reviewed (approved):
LGTM! @zou3519 could you please check that Dynamo changes are in the right places? Thanks!

davidberard98 committed:
Update on "[user triton] triton_kernel_wrap support for new host-side TMA API"


This adds support for user-defined triton kernels using TensorDescriptor.from_tensor into triton_kernel_wrap: i.e. storing metadata about the TMA descriptors and doing mutation analysis.

Major changes:
* TMADescriptorMetadata has changed: previously it was a dict[str, tuple[list[int], list[int], int]]. But now there are two metadata formats: one for experimental API and one for stable API. Now the metadata format is dict[str, tuple[str, tuple[...]]], where tuple[...] is tuple[list[int], list[int], int] for experimental and tuple[list[int],] for stable API. And then most handling of the metadata has to be branched based on whether the metadata represents a stable or experimental TMA descriptor
* mutation analysis: unlike experimental TMA (where the mutation analysis / ttir analysis pretends that the TMA descriptor is actually just a tensor), we need to construct an actual TMA descriptor before getting the Triton frontend to create the TTIR (otherwise assertions fail). A TensorDescriptor (i.e. stable TMA API descriptor) passed into a python triton kernel actually turns into 1 + 2*N parameters in the TTIR (for a rank-N tensor), so the arg list also needs to be patched for this reason (in generate_ttir)
* mutation analysis: now we also need to pass tma_descriptor_metadata into the mutation analysis, in order to create the TMA descriptors that are passed into the frontend code (ie. the previous point). This is why all the mutation tests are modified with an extra return value (the tma_descriptor_metadata)

Inductor is not modified (Inductor just errors out if you use a stable API tma descriptor). This will be the next PR.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 reviewed (commented):
seems reasonable to me
